### PR TITLE
llists tests using KUnit

### PR DESCRIPTION
**Pull request by tutuzeraa**

### CODE REVIEW

File: lib/Kconfig [hunk 1]
```diff
@@ -776,3 +776,8 @@ config POLYNOMIAL
 
 config FIRMWARE_TABLE
 	bool
+
+config TEST_LLIST
```
Comment (diegodvv) [inline lib/Kconfig]: A config dos testes do KUnit costumam ter a forma `<STRUCT>_KUNIT_TEST` como `LIST_KUNIT_TEST`
Nesse caso ficaria `LLIST_KUNIT_TEST`

File: lib/Makefile [hunk 2]
```diff
@@ -84,6 +84,7 @@ obj-$(CONFIG_TEST_STATIC_KEYS) += test_static_key_base.o
 obj-$(CONFIG_TEST_DYNAMIC_DEBUG) += test_dynamic_debug.o
 obj-$(CONFIG_TEST_PRINTF) += test_printf.o
 obj-$(CONFIG_TEST_SCANF) += test_scanf.o
+obj-$(CONFIG_TEST_LLIST) += llist_kunit.o
```
Comment (diegodvv) [inline lib/Makefile]: Acho melhor adicionar a linha perto das linhas dos outros testes de KUnit, na linha 371 desse Makefile

File: lib/llist_kunit.c [hunk 3]
```diff
@@ -0,0 +1,366 @@
+#include <kunit/test.h>
```
Comment (diegodvv) [inline lib/llist_kunit.c:8]: Geralmente o topo dos arquivos de testes do KUnit tem um comentário sobre o que se trata e com o nome do autor (pelo menos list-test.c tinha)

File: lib/llist_kunit.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    KUNIT_EXPECT_PTR_EQ(test, &test_struct, llist_entry(&(test_struct.node), struct llist_test_struct, node));
+}
+
+static void llist_test_member_address_is_nonnull(struct kunit *test)
+{
+    struct llist_test_struct test_struct;
+    
+    KUNIT_EXPECT_TRUE(test, member_address_is_nonnull(&test_struct, node)); 
+
+    /* TODO conseguir testar essa bagaca*/
```
Comment (diegodvv) [inline lib/llist_kunit.c]: Teste incompleto com TODO

File: lib/llist_kunit.c [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    llist_add(&b, &llist);
+    
+    llist_del_first_this(&llist, &a);
+    
+    /* before: [List] -> b -> a */
+    KUNIT_EXPECT_FALSE(test, llist_del_first_this(&llist, &a)); // it should remove only if is the first node in the llist
+                                                                
+    KUNIT_EXPECT_TRUE(test, llist_del_first_this(&llist, &b));
+
+    /* shoulde be [List] -> a */
```
Comment (diegodvv) [inline lib/llist_kunit.c]: Typo

File: lib/llist_kunit.c [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    KUNIT_CASE(llist_test_for_each),
+    KUNIT_CASE(llist_test_for_each_safe),
+    KUNIT_CASE(llist_test_for_each_entry),
+    KUNIT_CASE(llist_test_for_each_entry_safe),
+    KUNIT_CASE(llist_test_reverse_order),
+    {}
+};
+
+static struct kunit_suite llist_test_suite = {
+    .name = "lib-llist",
```
Comment (diegodvv) [inline lib/llist_kunit.c]: Nos outros testes deixam o nome da suite só como o nome da estrutura, ficaria
`.name = "llist",`

File: lib/Kconfig [hunk 7]
```diff
@@ -776,3 +776,8 @@ config POLYNOMIAL
...
+    tristate "KUnit tests for lib/llist" if !KUNIT_ALL_TESTS
+    default KUNIT_ALL_TESTS
+    depends on KUNIT
```
Comment (diegodvv) [inline lib/Kconfig]: Faltou adicionar o help (descrição da config). Eu fiz quase um copy paste da descrição do LIST_KUNIT_TEST

File: lib/Kconfig.debug [hunk 8]
```diff
@@ -2814,6 +2814,16 @@ config SIPHASH_KUNIT_TEST
...
+config LLIST_KUNIT_TEST
+    tristate "KUnit tests for lib/llist" if !KUNIT_ALL_TESTS
+    default KUNIT_ALL_TESTS
+    depends on KUNIT
+    help
+        This option builds the "llist_kunit" test module that
+        helps to verify it correctness. 
+
+        If unsure, say N.
+
```
Comment (nukelet) [inline lib/Kconfig.debug]: these whitespaces should be tabs instead

File: lib/llist_kunit.c [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+static void llist_test_member_address_is_nonnull(struct kunit *test)
+{
+	struct llist_test_struct test_struct;
+
+	KUNIT_EXPECT_TRUE(test, member_address_is_nonnull(&test_struct, node));
+
+	/* TODO conseguir testar essa bagaca*/
+	//test_struct_ptr = NULL;
+	//KUNIT_EXPECT_FALSE(test, member_address_is_nonnull(test_struct_ptr, node));
```
Comment (nukelet) [inline lib/llist_kunit.c]: dont want to be sending this to linus :-)

File: lib/llist_kunit.c [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+	init_llist_head(llist5);
+
+	KUNIT_EXPECT_TRUE(test, llist_empty(&llist1));
+	KUNIT_EXPECT_TRUE(test, llist_empty(&llist2));
+	KUNIT_EXPECT_TRUE(test, llist_empty(&llist3));
+	KUNIT_EXPECT_TRUE(test, llist_empty(llist4));
+	KUNIT_EXPECT_TRUE(test, llist_empty(llist5));
+
+	kfree(llist4);
+	kfree(llist5);
```
Comment (nukelet) [inline lib/llist_kunit.c]: i think doing this sequentially makes it easier to understand that you're testing whether all different initialization methods leave the list empty:

```c
	struct llist_head llist1 = LLIST_HEAD_INIT(llist1);
	KUNIT_EXPECT_TRUE(test, llist_empty(&llist1));

	LLIST_HEAD(llist2);
	KUNIT_EXPECT_TRUE(test, llist_empty(&llist2));

	struct llist_head llist3;
	init_llist_head(&llist3);
	KUNIT_EXPECT_TRUE(test, llist_empty(&llist3));

	struct llist_head *llist4;
	llist4 = kzalloc(sizeof(*llist4), GFP_KERNEL | __GFP_NOFAIL);
	init_llist_head(llist4);
	KUNIT_EXPECT_TRUE(test, llist_empty(llist4));
	kfree(llist4);

	struct llist_head *llist5;
	llist5 = kmalloc(sizeof(*llist5), GFP_KERNEL | __GFP_NOFAIL);
	memset(llist5, 0xFF, sizeof(*llist5));
	init_llist_head(llist5);
	KUNIT_EXPECT_TRUE(test, llist_empty(llist5));
	kfree(llist5);
```

File: lib/llist_kunit.c [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+
+#include <kunit/test.h>
+#include <linux/llist.h>
+
+struct llist_test_struct {
+	int data;
+	struct llist_node node;
+};
+
+static void llist_test_init_llist(struct kunit *test)
```
Comment (nukelet) [inline lib/llist_kunit.c:18]: maybe name this `llist_test_init_is_empty`?

File: lib/llist_kunit.c [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+	KUNIT_EXPECT_PTR_EQ(test, a.next, &a);
+}
+
+static void llist_test_llist_entry(struct kunit *test)
+{
+	struct llist_test_struct test_struct;
+
+	KUNIT_EXPECT_PTR_EQ(test, &test_struct,
+			   llist_entry(&(test_struct.node),
+					struct llist_test_struct, node));
```
Comment (nukelet) [inline lib/llist_kunit.c]: i would make this a little bit more explicit:

```c
static void llist_test_llist_entry(struct kunit *test)
{
        struct llist_test_struct test_struct;

        struct llist *list = &test_struct.node;
        KUNIT_EXPECT_PTR_EQ(test, &test_struct,
                        llist_entry(list, struct llist_test_struct, node));
}      
```

File: lib/llist_kunit.c [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+
+static void llist_test_add(struct kunit *test)
+{
+	struct llist_node a, b;
+	LLIST_HEAD(llist);
+
+	init_llist_node(&a);
+	init_llist_node(&b);
+
+	/*The first assertion must be true, given that llist is empty */
```
Comment (nukelet) [inline lib/llist_kunit.c]: nit: missing whitespace between `/*` and `The`

File: lib/llist_kunit.c [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+	llist_add(&b, &llist);
+
+	/* before: [List] -> b -> a */
+	c = llist_del_first(&llist);
+
+	/* should be [List] -> a */
+	KUNIT_EXPECT_PTR_EQ(test, llist.first, &a);
+
+	/* del must return a pointer to llist_node b 
+     * the returned pointer must be marked on list*/
```
Comment (nukelet) [inline lib/llist_kunit.c]: nit: formatting

File: lib/llist_kunit.c [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+{
+	struct llist_node a;
+	LLIST_HEAD(llist);
+
+	llist_add(&a, &llist);
+
+	struct llist_node *b = llist_del_first_init(&llist);
+
+	/* should be [List] 
+     * the returned pointer must be marked out of the list */
```
Comment (nukelet) [inline lib/llist_kunit.c]: nit: formatting

File: lib/llist_kunit.c [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+	llist_add(&b, &llist);
+
+	llist_del_first_this(&llist, &a);
+
+	/* before: [List] -> b -> a */
+	KUNIT_EXPECT_FALSE(
+		test,
+		llist_del_first_this(
+			&llist,
+			&a)); // it should remove only if is the first node in the llist
```
Comment (nukelet) [inline lib/llist_kunit.c]: formatting nit: this should be a little better
```c
        // should be removed only if it is the first node in the list
	KUNIT_EXPECT_FALSE(test, llist_del_first_this(&llist, &a));
```

File: lib/llist_kunit.c [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+	/* before [List] -> entries[0] -> entries[1] -> entries[2] */
+
+	llist_for_each(pos, llist.first) {
+		KUNIT_EXPECT_PTR_EQ(test, pos, &entries[i]);
+		i++;
+	}
+
+	i = 0;
+
+	/* transversing deleted node */
```
Comment (nukelet) [inline lib/llist_kunit.c]: nit: s/transversing/traversing, s/node/nodes

File: lib/llist_kunit.c [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+	KUNIT_EXPECT_NULL(test, llist_del_all(&empty_llist));
+
+	llist_del_all(&llist);
+
+	KUNIT_EXPECT_TRUE(test, llist_empty(&llist));
+}
+
+static void llist_test_for_each(struct kunit *test)
+{
+	struct llist_node entries[3], *pos, *deleted_nodes;
```
Comment (nukelet) [inline lib/llist_kunit.c]: `entries` needs to be initialized!
```c
struct llist_node entries[3] = { 0 };
struct llist_node *pos, *deleted_nodes;
```

File: lib/llist_kunit.c [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+		KUNIT_EXPECT_PTR_EQ(test, pos, &entries[i]);
+		i++;
+	}
+
+	KUNIT_EXPECT_EQ(test, 3, i);
+}
+
+static void llist_test_for_each_safe(struct kunit *test)
+{
+	struct llist_node entries[3], *pos, *n;
```
Comment (nukelet) [inline lib/llist_kunit.c]: same as before, consider initializing `entries` to 0

**Review by nukelet — CHANGES_REQUESTED** [review]
great work on the tests! just a few observations

File: lib/llist_kunit.c (hunk 11 above)
Comment (nukelet) [reply lib/llist_kunit.c:18]: from meeting: add comment explaining this tests whether the lists are empty when being initialized

File: lib/llist_kunit.c [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+#include <linux/llist.h>
+
+struct llist_test_struct {
+	int data;
+	struct llist_node node;
+};
+
+static void llist_test_init_llist(struct kunit *test)
+{
+	struct llist_head llist1 = LLIST_HEAD_INIT(llist1);
```
Comment (andrealmeid) [inline lib/llist_kunit.c:21]: `struct llist_head llist1 = LLIST_HEAD_INIT(llist1), llist2, llist3, llist4;`

File: lib/llist_kunit.c (hunk 11 above)
Comment (nukelet) [reply lib/llist_kunit.c:18]: from meeting: make sure to run checkpatch.pl on this

File: lib/llist_kunit.c [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+	init_llist_node(&a);
+
+	KUNIT_EXPECT_PTR_EQ(test, a.next, &a);
+}
+
+static void llist_test_llist_entry(struct kunit *test)
+{
+	struct llist_test_struct test_struct;
+
+    struct llist_node *list = &test_struct.node;
```
Comment (andrealmeid) [inline lib/llist_kunit.c]: tab

File: lib/llist_kunit.c [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+	KUNIT_EXPECT_PTR_EQ(test, &test_struct,
+			   llist_entry(list, struct llist_test_struct, node));
```
Comment (andrealmeid) [inline lib/llist_kunit.c]: use a aux var here

```
aux = llist_entry(list, struct llist_test_struct, node);
KUNIT_EXPECT_PTR_EQ(test, &test_struct, aux);
```

Comment (andrealmeid) [reply lib/llist_kunit.c]:
```
	KUNIT_EXPECT_PTR_EQ(test, &test_struct,
			    llist_entry(list, struct llist_test_struct, node));
```

**Review by nukelet — COMMENTED** [review]
a few comments

File: lib/llist_kunit.c [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+	llist_add(&a, &llist);
+	llist_add(&b, &llist);
+
+	/* before: [List] -> b -> a */
+	c = llist_del_first(&llist);
+
+	/* should be [List] -> a */
+	KUNIT_EXPECT_PTR_EQ(test, llist.first, &a);
+
+	/* del must return a pointer to llist_node b
```
Comment (andrealmeid) [inline lib/llist_kunit.c]:
```
/*
 * abcd
 */
```

File: lib/llist_kunit.c [hunk 24]
```diff
@@ -0,0 +1,363 @@
...
+	int i = 0;
+
+	llist_add(&entries[2], &llist);
+	llist_add(&entries[1], &llist);
+	llist_add(&entries[0], &llist);
+
+	/* before [List] -> entries[0] -> entries[1] -> entries[2] */
+
+	llist_for_each(pos, llist.first) {
+		KUNIT_EXPECT_PTR_EQ(test, pos, &entries[i]);
```
Comment (andrealmeid) [inline lib/llist_kunit.c]: `KUNIT_EXPECT_PTR_EQ(test, pos, &entries[i++]);`

File: lib/llist_kunit.c [hunk 25]
```diff
@@ -0,0 +1,363 @@
...
+	KUNIT_EXPECT_NULL(test, llist_del_all(&empty_llist));
+
+	llist_del_all(&llist);
+
+	KUNIT_EXPECT_TRUE(test, llist_empty(&llist));
+}
+
+static void llist_test_for_each(struct kunit *test)
+{
+	struct llist_node entries[3] = { 0 };
```
Comment (andrealmeid) [inline lib/llist_kunit.c]:
```c
#DEFINE ENTRIES_SIZE 3

struct llist_node entries[ENTRIES_SIZE] = { 0 };


KUNIT_EXPECT_EQ(test, ENTRIES_SIZE, i);
```

File: lib/llist_kunit.c [hunk 26]
```diff
@@ -0,0 +1,363 @@
...
+	struct llist_node *pos, *deleted_nodes;
+	LLIST_HEAD(llist);
+	int i = 0;
+
+	llist_add(&entries[2], &llist);
```
Comment (andrealmeid) [inline lib/llist_kunit.c]: use a for loop here

File: lib/llist_kunit.c [hunk 27]
```diff
@@ -0,0 +1,363 @@
...
+	llist_add(&entries[1], &llist);
+	llist_add(&entries[0], &llist);
+
+	/* before [List] -> entries[0] -> entries[1] -> entries[2] */
+
+	llist_for_each(pos, llist.first) {
+		KUNIT_EXPECT_PTR_EQ(test, pos, &entries[i]);
+		i++;
+	}
```
Comment (andrealmeid) [inline lib/llist_kunit.c:234]: add `KUNIT_EXPECT_EQ(test, 3, i);` here

File: lib/llist_kunit.c [hunk 28]
```diff
@@ -0,0 +1,363 @@
...
+	int i = 0;
+
+	llist_add(&entries[0], &llist);
+	llist_add(&entries[1], &llist);
+	llist_add(&entries[2], &llist);
+
+	/* before [List] -> entries[2] -> entries[1] -> entries[0] */
+	reversed_llist = llist_reverse_order(llist_del_first(&llist));
+
+	/* should be [List] -> entries[0] -> entries[1] -> entrires[0] */
```
Comment (andrealmeid) [inline lib/llist_kunit.c]:
```
entries[0] -> entries[1] -> entries[2]
```
